src/main/java/com/gtnewhorizon/gtnhlib/capability/CapabilityUtil.java

```diff
@@ -0,0 +1,111 @@
+ * traditional interface implementations.
+ * <p>
+ * Example usage:
+ * <pre>{@code
+ *     // Retrieve capability from a TileEntity
+ *     MyInterface impl = CapabilityUtil.getCapability(tileEntity, MY_CAPABILITY, ForgeDirection.NORTH);
+ *
+ *     // Retrieve capability from an ItemStack
+ *     MyInterface impl = CapabilityUtil.getCapability(itemStack, MY_CAPABILITY);
+ *
+ *     // Retrieve capability from an Entity
+ *     MyInterface impl = CapabilityUtil.getCapability(entity, MY_CAPABILITY);
+ * }</pre>
+ */
+// spotless:on
+public final class CapabilityUtil {
+
+    /**
+     * Retrieves a capability from the given TileEntity.
+     *
+     * @param tileEntity The TileEntity to query.
+     * @param capability The {@link Capability} being requested.
+     * @param side       The side of the TileEntity being queried.
+     * @param <T>        The type of the capability interface.
+     * @return The capability implementation, or null if not available.
+     */
+    public static <T> T getCapability(@Nullable TileEntity tileEntity, @NotNull Capability<T> capability,
+            @NotNull ForgeDirection side) {
+        return getCapability((Object) tileEntity, capability, side);
+    }
+
+    /**
+     * Retrieves a capability from the given TileEntity without specifying a side.
+     *
+     * @param tileEntity The TileEntity to query.
+     * @param capability The {@link Capability} being requested.
+     * @param <T>        The type of the capability interface.
+     * @return The capability implementation, or null if not available.
+     */
+    public static <T> T getCapability(@Nullable TileEntity tileEntity, @NotNull Capability<T> capability) {
+        return getCapability((Object) tileEntity, capability, ForgeDirection.UNKNOWN);
+    }
+
+    /**
+     * Retrieves a capability from the given ItemStack's Item.
+     *
+     * @param itemStack  The ItemStack to query.
+     * @param capability The {@link Capability} being requested.
+     * @param <T>        The type of the capability interface.
+     * @return The capability implementation, or null if not available.
+     */
+    public static <T> T getCapability(@Nullable ItemStack itemStack, @NotNull Capability<T> capability) {
+        if (itemStack == null) {
+            return null;
+        }
+        Item item = itemStack.getItem();
+        return getCapability(item, capability, ForgeDirection.UNKNOWN);
+    }
+
+    /**
+     * Retrieves a capability from the given Entity.
+     *
+     * @param entity     The Entity to query.
+     * @param capability The {@link Capability} being requested.
+     * @param <T>        The type of the capability interface.
+     * @return The capability implementation, or null if not available.
+     */
+    public static <T> T getCapability(@Nullable Entity entity, @NotNull Capability<T> capability) {
+        return getCapability(entity, capability, ForgeDirection.UNKNOWN);
+    }
```

Comment:
Why does this only accept an Entity while the other override accepts an Object & a Side? This means if I want to get a capability of a non-entity, a side is required?

Reply:
Method accepting `Object` is set to private. Actually anything can work with capability, but I want to ensure type safety and listed things that one might want to use capability in MC modding context.